flink-runtime/src/main/java/org/apache/flink/runtime/resourcemanager/slotmanager/ResourceAllocationStrategy.java

```diff
@@ -46,4 +46,14 @@ ResourceAllocationResult tryFulfillRequirements(
             Map<JobID, Collection<ResourceRequirement>> missingResources,
             TaskManagerResourceInfoProvider taskManagerResourceInfoProvider,
             BlockedTaskManagerChecker blockedTaskManagerChecker);
+
+    /**
+     * Try to fulfill the missing redundant requirements.
+     *
+     * @param missingRedundantResource redundant resources that are not yet fulfilled
+     * @return the new pending task managers that need to allocated to fulfill redundant
+     *     requirements.
+     */
+    Collection<PendingTaskManager> tryFulfillRedundantRequirements(
+            RequiredRedundantResource missingRedundantResource);
```

Comment:
I wonder if it make sense to let the strategy decide whether they should allocate for redundant workers. To be specific, we construct the `DefaultResourceAllocationStrategy.java` with an argument `redundantTaskManagerNum`. Then `ResourceAllocationResult` returned from the strategy should already include redundant workers.